### DISCUSSION TITLE
RS485 Driver Enable not being enabled [STM32H7]

### DISCUSSION THREAD

**Discussion by xavierpacheco-eaton** [zephyrproject-rtos/zephyr#79007] [Q&A]
When configuring UART in the devicetree I'm enabling the DE pin as follows:

```
 &uart5 {
    pinctrl-0 = <&uart5_tx_pb13 &uart5_rx_pb12 &uart5_de_pc8>;
    current-speed = <115200>;
    status = "okay";
    pinctrl-names="default";
    de-enable; 
};
```

However, this doesn't set the USART_CR3_DEM as the macro HAS_DRIVER_ENABLE is 0. I'm using Zephyr 3.6.0 and STM32CubeH7 v1.11.1.

Any pointers whether I'm missing something or could this be an implementation issue?

**Comment by KozhinovAlexander** [comment]
Please add more infomration:

1. HW/Board name - if not custom
2. The exact model/version of stm32h7 SoC

Double-check if clocks and corresponding peripherals are enabled. Sometimes it may be necessary to take a look into registers after startup.

**Reply by xavierpacheco-eaton** [reply]
It's a custom board using the STM32H745XI. The UART is being enabled, except the DE feature. I enabled it manually inside uart_stm32.c and wondering by it's not working through the binding.

**Reply by KozhinovAlexander** [reply]
> It's a custom board using the STM32H745XI. The UART is being enabled, except the DE feature.

To be sure uart5 is enabled, you shall use debugger on running board and see if corresponding registeres (please refer to reference manual here) are set to proper state. 

> I enabled it manually inside uart_stm32.c and wondering by it's not working through the binding. 

uart is a device by Zephyr's definiteon and shall be handled like any device. 
In general it is possible to enable any device from any .c/.h file, but it is not recommended.

In the first place any device shall be enabled by setting `status = "okay"` from device tree. Afterwards it can be accessed by device tree API: https://docs.zephyrproject.org/apidoc/latest/group__device__model.html#details

I would suggest you start with clean main branch and build UART echo application: https://docs.zephyrproject.org/latest/samples/drivers/uart/echo_bot/README.html

By default on my nucleo_h745zi_q board this application runs over usart3, defined in this line as zephyr_shell_uart alias: https://github.com/zephyrproject-rtos/zephyr/blob/c710f8892beef3e486dded2f0ad4394e4a4d9aa0/boards/st/nucleo_h745zi_q/nucleo_h745zi_q_stm32h745xx_m7.dts#L25

You may replace it with your uart5, but do not forget to set `status = "okay"`.

Of course you need then something like uart-usb adapter to be able to see console output.

This would be a great test for custom board. Sometimes it could be possible that there are problems with layout, routing, production defects.
After verifying the sample and uart5 working fine, you may simply put uart5 in the line: https://github.com/zephyrproject-rtos/zephyr/blob/c710f8892beef3e486dded2f0ad4394e4a4d9aa0/samples/drivers/uart/echo_bot/src/main.c#L14

... or similarly build a piece of code for your purposes based on above sample.

If after all tries uart5 is still not working, but debugger shows proper bits set in all the registers and clocks are also enabled, your only choice would be an oscilloscope or logic analyzer connected to pins choosen by uart5. But this is not a scope of zephyr RTOS, instead a classical debugging technique for any HW.

UPD: DE pin shall/could be checked after above steps are fine.

**Reply by KozhinovAlexander** [reply]
BTW: Did you enabled CONFIG_SERIAL=y in your prj.conf?

**Reply by xavierpacheco-eaton** [reply]
UART5 works fine. The issue is the DE feature. I stepped through the code and indeed USART_CR3_DEM is not being set.

**Reply by KozhinovAlexander** [reply]
With changes from patch:

```
diff --git a/boards/st/nucleo_h745zi_q/nucleo_h745zi_q_stm32h745xx_m7.dts b/boards/st/nucleo_h745zi_q/nucleo_h745zi_q_stm32h745xx_m7.dts
index bce218ae37d..177993994d5 100644
--- a/boards/st/nucleo_h745zi_q/nucleo_h745zi_q_stm32h745xx_m7.dts
+++ b/boards/st/nucleo_h745zi_q/nucleo_h745zi_q_stm32h745xx_m7.dts
@@ -41,6 +41,7 @@
 		led0 = &green_led;
 		pwm-led0 = &red_pwm_led;
 		sw0 = &user_button;
+		rs485-uart = &uart5;
 	};
 };
 
@@ -80,6 +81,15 @@
 	status = "okay";
 };
 
+&uart5 {
+	pinctrl-0 = <&uart5_tx_pb13 &uart5_rx_pb12 &uart5_de_pc8>;
+	current-speed = <115200>;
+	status = "okay";
+	pinctrl-names="default";
+	de-enable;
+	// /delete-property/ fifo-enable;
+};
+
 &rtc {
 	clocks = <&rcc STM32_CLOCK_BUS_APB4 0x00010000>,
 		 <&rcc STM32_SRC_LSI RTC_SEL(2)>;
@@ -105,7 +115,7 @@
 };
 
 &mac {
-	status = "okay";
+	status = "disabled";
 	pinctrl-0 = <&eth_ref_clk_pa1
 		     &eth_crs_dv_pa7
 		     &eth_rxd0_pc4
diff --git a/samples/drivers/uart/echo_bot/src/main.c b/samples/drivers/uart/echo_bot/src/main.c
index 549c74642c5..9b9433e4567 100644
--- a/samples/drivers/uart/echo_bot/src/main.c
+++ b/samples/drivers/uart/echo_bot/src/main.c
@@ -11,7 +11,7 @@
 #include <string.h>
 
 /* change this to any other UART peripheral if desired */
-#define UART_DEVICE_NODE DT_CHOSEN(zephyr_shell_uart)
+#define UART_DEVICE_NODE DT_ALIAS(rs485_uart)
 
 #define MSG_SIZE 32
 
@@ -79,29 +79,31 @@ int main(void)
 		return 0;
 	}
 
-	/* configure interrupt and callback to receive data */
-	int ret = uart_irq_callback_user_data_set(uart_dev, serial_cb, NULL);
-
-	if (ret < 0) {
-		if (ret == -ENOTSUP) {
-			printk("Interrupt-driven UART API support not enabled\n");
-		} else if (ret == -ENOSYS) {
-			printk("UART device does not support interrupt-driven API\n");
-		} else {
-			printk("Error setting UART callback: %d\n", ret);
-		}
-		return 0;
-	}
-	uart_irq_rx_enable(uart_dev);
-
-	print_uart("Hello! I'm your echo bot.\r\n");
-	print_uart("Tell me something and press enter:\r\n");
-
-	/* indefinitely wait for input from the user */
-	while (k_msgq_get(&uart_msgq, &tx_buf, K_FOREVER) == 0) {
-		print_uart("Echo: ");
-		print_uart(tx_buf);
-		print_uart("\r\n");
-	}
+	while(true);  // put a beak-point here
+
+	// /* configure interrupt and callback to receive data */
+	// int ret = uart_irq_callback_user_data_set(uart_dev, serial_cb, NULL);
+	//
+	// if (ret < 0) {
+	// 	if (ret == -ENOTSUP) {
+	// 		printk("Interrupt-driven UART API support not enabled\n");
+	// 	} else if (ret == -ENOSYS) {
+	// 		printk("UART device does not support interrupt-driven API\n");
+	// 	} else {
+	// 		printk("Error setting UART callback: %d\n", ret);
+	// 	}
+	// 	return 0;
+	// }
+	// uart_irq_rx_enable(uart_dev);
+	//
+	// print_uart("Hello! I'm your echo bot.\r\n");
+	// print_uart("Tell me something and press enter:\r\n");
+	//
+	// /* indefinitely wait for input from the user */
+	// while (k_msgq_get(&uart_msgq, &tx_buf, K_FOREVER) == 0) {
+	// 	print_uart("Echo: ");
+	// 	print_uart(tx_buf);
+	// 	print_uart("\r\n");
+	// }
 	return 0;
 }
```
I see DEM for UART5 enabled:
![grafik](https://github.com/user-attachments/assets/b5b759b0-1851-461d-9af8-366f9f18ce47)

**Comment by ndrs-pst** [comment]
@xavierpacheco-eaton The only way `HAS_DRIVER_ENABLE` is 0 is when there is no `USART_CR3_DEM`, implying that the peripheral does not support the DE feature. If you can upload `zephyr.dts`, it would help a lot.

https://github.com/zephyrproject-rtos/zephyr/blob/097ed18842c5360795b437c0e53131e056917c10/drivers/serial/uart_stm32.c#L57-L61

**Reply by xavierpacheco-eaton** [reply]
The UART5 of the STM32H745XI supports the DE feature so it should be enabled.

**Reply by erwango** [reply]
`USART_CR3_DEM` is defined in https://raw.githubusercontent.com/zephyrproject-rtos/hal_stm32/refs/heads/main/stm32cube/stm32h7xx/soc/stm32h745xx.h so I don't see how definition could be missing... This is strange indeed
